File: watch/app/src/main/java/com/gymapp/watch/engine/SessionEngine.kt
package com.gymapp.watch.engine

import com.gymapp.watch.data.model.PlanExerciseEntry
import com.gymapp.watch.data.model.SerieEntry
import com.gymapp.watch.data.model.SessionExercise
import com.gymapp.watch.data.model.WorkoutPlan
import com.gymapp.watch.data.model.WorkoutSession
import java.util.UUID
import kotlin.math.roundToLong

/**
 * Porting 1:1 di src/workout/sessionEngine.js. Stessa logica di ordinamento e
 * posticipa richiesta esplicitamente dal documento dei flussi utente (F2.3):
 * "Stessa logica da replicare sull'app watch".
 *
 * Stile funzionale: ogni funzione riceve una WorkoutSession e ne ritorna una nuova
 * (nessuna mutazione in place), cosi' la UI (StateFlow) puo' confrontare vecchio/nuovo
 * stato con equals() strutturale delle data class.
 */
object SessionEngine {

    private fun now(): Long = System.currentTimeMillis()

    /** Ordine iniziale: raggruppa per categoria mantenendo l'ordine di prima apparizione nella scheda */
    fun buildOrder(exercises: List<PlanExerciseEntry>): List<String> {
        val catOrder = mutableListOf<String>()
        for (e in exercises) if (e.category !in catOrder) catOrder.add(e.category)
        val order = mutableListOf<String>()
        for (cat in catOrder) {
            for (e in exercises) if (e.category == cat) order.add(e.key)
        }
        return order
    }

    /** Crea la sessione: snapshot della scheda + stato di esecuzione iniziale */
    fun createSession(plan: WorkoutPlan, restDefaultSec: Int): WorkoutSession {
        val exercises = plan.exercises.map { e ->
            SessionExercise(
                key = e.key,
                refType = e.refType,
                refId = e.refId,
                name = e.name,
                category = e.category,
                image = e.image,
                description = e.description,
                mode = e.mode,
                durationSec = e.durationSec,
                sets = e.sets,
                reps = e.reps,
                hasWeight = e.hasWeight,
                weightKg = e.weightKg,
                postponeCount = 0,
                skipped = false,
                startedAt = null,
                endedAt = null,
                note = "",
                series = List(e.sets) {
                    SerieEntry(
                        done = false,
                        startedAt = null,
                        doneAt = null,
                        actualReps = e.reps,
                        actualWeightKg = if (e.hasWeight) e.weightKg else null,
                        restSec = null,
                    )
                },
            )
        }
        return WorkoutSession(
            id = "s-${now()}-${UUID.randomUUID().toString().take(6)}",
            planId = plan.id,
            planName = plan.name,
            planColor = plan.color,
            origine = "watch",
            restDefaultSec = restDefaultSec,
            startedAt = now(),
            endedAt = null,
            status = "active",
            pausedMs = 0,
            pauseStartedAt = null,
            exercises = exercises,
            queue = buildOrder(plan.exercises),
        )
    }

    fun getExercise(session: WorkoutSession, key: String): SessionExercise? =
        session.exercises.find { it.key == key }

    fun currentExercise(session: WorkoutSession): SessionExercise? =
        session.queue.firstOrNull()?.let { getExercise(session, it) }

    /** Indice della prossima serie non fatta, o -1 */
    fun nextUndoneSerie(exercise: SessionExercise): Int =
        exercise.series.indexOfFirst { !it.done }

    private fun WorkoutSession.updateExercise(key: String, transform: (SessionExercise) -> SessionExercise): WorkoutSession =
        copy(exercises = exercises.map { if (it.key == key) transform(it) else it })

    private fun SessionExercise.updateSerie(idx: Int, transform: (SerieEntry) -> SerieEntry): SessionExercise =
        copy(series = series.mapIndexed { i, s -> if (i == idx) transform(s) else s })

    /** Avvia una serie (stato "in corso") */
    fun startSerie(session: WorkoutSession, key: String, serieIdx: Int): WorkoutSession =
        session.updateExercise(key) { ex ->
            ex.copy(
                startedAt = ex.startedAt ?: now(),
            ).updateSerie(serieIdx) { it.copy(startedAt = now()) }
        }

    /** Serie fatta: registra effettivi e timestamp. */
    fun markSerieDone(
        session: WorkoutSession,
        key: String,
        serieIdx: Int,
        actualReps: Int? = null,
        actualWeightKg: Double? = null,
        weightProvided: Boolean = false,
    ): WorkoutSession = session.updateExercise(key) { ex ->
        val withStart = if (ex.startedAt == null) ex.copy(startedAt = now()) else ex
        val updated = withStart.updateSerie(serieIdx) { serie ->
            serie.copy(
                done = true,
                doneAt = now(),
                actualReps = actualReps ?: serie.actualReps,
                actualWeightKg = if (weightProvided) actualWeightKg else serie.actualWeightKg,
            )
        }
        if (updated.series.all { it.done }) updated.copy(endedAt = now()) else updated
    }.let { s ->
        val ex = getExercise(s, key)!!
        if (ex.series.all { it.done }) s.copy(queue = s.queue.filterNot { it == key }) else s
    }

    /** Correzione a posteriori di reps/peso di una serie */
    fun updateSerieActuals(
        session: WorkoutSession,
        key: String,
        serieIdx: Int,
        actualReps: Int? = null,
        actualWeightKg: Double? = null,
        weightProvided: Boolean = false,
    ): WorkoutSession = session.updateExercise(key) { ex ->
        ex.updateSerie(serieIdx) { serie ->
            serie.copy(
                actualReps = actualReps ?: serie.actualReps,
                actualWeightKg = if (weightProvided) actualWeightKg else serie.actualWeightKg,
            )
        }
    }

    /** Registra la durata effettiva del recupero dopo una serie */
    fun recordRest(session: WorkoutSession, key: String, serieIdx: Int, restSec: Int): WorkoutSession =
        session.updateExercise(key) { ex -> ex.updateSerie(serieIdx) { it.copy(restSec = restSec) } }

    /**
     * Posticipa l'esercizio corrente (F2.3):
     * 1a volta -> in coda alla propria categoria
     * 2a volta -> in fondo a tutto l'allenamento
     * 3a volta -> marcato saltato per questa sessione
     */
    fun postponeCurrent(session: WorkoutSession): WorkoutSession {
        val key = session.queue.firstOrNull() ?: return session
        val ex = getExercise(session, key) ?: return session
        val newPostponeCount = ex.postponeCount + 1
        var newQueue = session.queue.drop(1)
        var skipped = false

        when (newPostponeCount) {
            1 -> {
                var insertAt = newQueue.size
                for (i in newQueue.indices.reversed()) {
                    val other = getExercise(session, newQueue[i])
                    if (other?.category == ex.category) {
                        insertAt = i + 1
                        break
                    }
                }
                newQueue = newQueue.toMutableList().apply { add(insertAt, key) }
            }
            2 -> {
                newQueue = newQueue + key
            }
            else -> {
                skipped = true
            }
        }

        return session
            .copy(queue = newQueue)
            .updateExercise(key) { it.copy(postponeCount = newPostponeCount, skipped = skipped || it.skipped) }
    }

    /** Salta definitivamente l'esercizio corrente */
    fun skipCurrent(session: WorkoutSession): WorkoutSession {
        val key = session.queue.firstOrNull() ?: return session
        return session
            .copy(queue = session.queue.drop(1))
            .updateExercise(key) { it.copy(skipped = true) }
    }

    /** Sposta un esercizio scelto manualmente in testa alla coda */
    fun jumpTo(session: WorkoutSession, key: String): WorkoutSession {
        if (key !in session.queue) return session
        return session.copy(queue = listOf(key) + session.queue.filterNot { it == key })
    }

    fun setNote(session: WorkoutSession, key: String, note: String): WorkoutSession =
        session.updateExercise(key) { it.copy(note = note) }

    fun togglePause(session: WorkoutSession): WorkoutSession =
        if (session.pauseStartedAt != null) {
            session.copy(
                pausedMs = session.pausedMs + (now() - session.pauseStartedAt!!),
                pauseStartedAt = null,
            )
        } else {
            session.copy(pauseStartedAt = now())
        }

    /** [at] permette di retrodatare la chiusura (watchdog anti-dimenticanza) */
    fun finishSession(session: WorkoutSession, at: Long = now()): WorkoutSession {
        var s = session
        if (s.pauseStartedAt != null) {
            s = s.copy(pausedMs = s.pausedMs + (at - s.pauseStartedAt!!).coerceAtLeast(0L), pauseStartedAt = null)
        }
        s = s.copy(endedAt = at)
        val allDone = s.exercises.all { it.skipped || it.series.all { serie -> serie.done } }
        val anyDone = s.exercises.any { it.series.any { serie -> serie.done } }
        s = s.copy(
            status = if (allDone || anyDone) "completed" else "aborted",
            completedFully = allDone,
        )
        return s
    }

    /** Durata attiva in ms (esclude le pause) */
    fun activeDuration(s: WorkoutSession, at: Long = now()): Long {
        val end = s.endedAt ?: at
        var paused = s.pausedMs
        if (s.pauseStartedAt != null) paused += end - s.pauseStartedAt!!
        return maxOf(0L, end - s.startedAt - paused)
    }

    data class SessionStats(
        val durationSec: Long,
        val totalSeries: Int,
        val doneSeries: Int,
        val doneExercises: Int,
        val totalExercises: Int,
        val skipped: Int,
        val volumeKg: Long,
        val avgRestSec: Int?,
        val restTargetSec: Int,
    )

    /** Statistiche di riepilogo, versione compatta per lo schermo del watch */
    fun computeStats(s: WorkoutSession): SessionStats {
        val totalSeries = s.exercises.sumOf { it.sets }
        val doneSeries = s.exercises.sumOf { e -> e.series.count { it.done } }
        val doneExercises = s.exercises.count { e -> e.series.all { it.done } }
        val skipped = s.exercises.count { it.skipped }

        val volumeKg = s.exercises.sumOf { e ->
            e.series.filter { it.done && it.actualWeightKg != null }
                .sumOf { (it.actualReps ?: 0) * (it.actualWeightKg ?: 0.0) }
        }

        val allRests = s.exercises.flatMap { e -> e.series.mapNotNull { it.restSec } }
        val avgRestSec = if (allRests.isNotEmpty()) (allRests.sum().toDouble() / allRests.size).roundToLong().toInt() else null

        return SessionStats(
            durationSec = activeDuration(s) / 1000,
            totalSeries = totalSeries,
            doneSeries = doneSeries,
            doneExercises = doneExercises,
            totalExercises = s.exercises.size,
            skipped = skipped,
            volumeKg = volumeKg.roundToLong(),
            avgRestSec = avgRestSec,
            restTargetSec = s.restDefaultSec,
        )
    }
}

File: src/workout/sessionEngine.js
/**
 * Motore della sessione di allenamento (logica pura, senza UI).
 *
 * Regole di ordinamento (dal documento dei flussi):
 * - esercizi raggruppati per categoria; ordine categorie = prima apparizione nella scheda
 * - Posticipa 1ª volta -> in coda alla propria categoria
 * - Posticipa 2ª volta -> in fondo all'allenamento
 * - Posticipa 3ª volta -> saltato per questa sessione
 */

const now = () => Date.now()

/** Ordine iniziale: raggruppa per categoria mantenendo l'ordine della scheda */
export function buildOrder(exercises) {
  const catOrder = []
  for (const e of exercises) if (!catOrder.includes(e.category)) catOrder.push(e.category)
  const order = []
  for (const cat of catOrder) {
    for (const e of exercises) if (e.category === cat) order.push(e.key)
  }
  return order
}

/** Crea la sessione (snapshot della scheda + stato runtime) */
export function createSession(plan, restDefaultSec) {
  const exercises = plan.exercises.map((e) => ({
    key: e.key,
    refType: e.refType,
    refId: e.refId,
    name: e.name,
    category: e.category,
    image: e.image || null,
    description: e.description || '',
    mode: e.mode || 'reps',
    durationSec: e.durationSec ?? null,
    sets: e.sets,
    reps: e.reps ?? null,
    hasWeight: e.hasWeight,
    weightKg: e.weightKg ?? null,
    postponeCount: 0,
    skipped: false,
    startedAt: null,
    endedAt: null,
    note: '',
    series: Array.from({ length: e.sets }, () => ({
      done: false,
      startedAt: null,
      doneAt: null,
      actualReps: e.reps ?? null,
      actualWeightKg: e.hasWeight ? e.weightKg : null,
      restSec: null,
    })),
  }))
  return {
    id: `s-${now()}-${Math.random().toString(36).slice(2, 7)}`,
    planId: plan.id,
    planName: plan.name,
    planColor: plan.color || null,
    origine: 'web',
    restDefaultSec,
    startedAt: now(),
    endedAt: null,
    status: 'active', // active | completed | aborted
    pausedMs: 0,
    pauseStartedAt: null,
    exercises,
    queue: buildOrder(plan.exercises), // chiavi degli esercizi ancora da fare
  }
}

export const getExercise = (s, key) => s.exercises.find((e) => e.key === key)
export const currentExercise = (s) => (s.queue.length ? getExercise(s, s.queue[0]) : null)

/** Avvia una serie (stato "in corso") */
export function startSerie(session, key, serieIdx) {
  const s = structuredClone(session)
  const ex = getExercise(s, key)
  if (!ex.startedAt) ex.startedAt = now()
  ex.series[serieIdx].startedAt = now()
  return s
}

/** Indice della prossima serie non fatta, o -1 */
export function nextUndoneSerie(exercise) {
  return exercise.series.findIndex((x) => !x.done)
}

/** Serie fatta: registra effettivi e timestamp. Ritorna la sessione aggiornata. */
export function markSerieDone(session, key, serieIdx, actuals = {}) {
  const s = structuredClone(session)
  const ex = getExercise(s, key)
  if (!ex.startedAt) ex.startedAt = now()
  const serie = ex.series[serieIdx]
  serie.done = true
  serie.doneAt = now()
  if (actuals.actualReps != null) serie.actualReps = actuals.actualReps
  if (actuals.actualWeightKg !== undefined) serie.actualWeightKg = actuals.actualWeightKg
  if (ex.series.every((x) => x.done)) {
    ex.endedAt = now()
    s.queue = s.queue.filter((k) => k !== key)
  }
  return s
}

/** Correzione a posteriori di reps/peso di una serie */
export function updateSerieActuals(session, key, serieIdx, actuals) {
  const s = structuredClone(session)
  const serie = getExercise(s, key).series[serieIdx]
  if (actuals.actualReps != null) serie.actualReps = actuals.actualReps
  if (actuals.actualWeightKg !== undefined) serie.actualWeightKg = actuals.actualWeightKg
  return s
}

/** Registra la durata effettiva del recupero dopo una serie */
export function recordRest(session, key, serieIdx, restSec) {
  const s = structuredClone(session)
  const ex = getExercise(s, key)
  if (ex && ex.series[serieIdx]) ex.series[serieIdx].restSec = restSec
  return s
}

/** Posticipa l'esercizio corrente secondo le regole */
export function postponeCurrent(session) {
  const s = structuredClone(session)
  const key = s.queue[0]
  const ex = getExercise(s, key)
  ex.postponeCount += 1
  s.queue = s.queue.slice(1)

  if (ex.postponeCount === 1) {
    // in coda alla propria categoria
    let insertAt = s.queue.length
    for (let i = s.queue.length - 1; i >= 0; i--) {
      if (getExercise(s, s.queue[i]).category === ex.category) { insertAt = i + 1; break }
    }
    s.queue.splice(insertAt, 0, key)
  } else if (ex.postponeCount === 2) {
    s.queue.push(key)
  } else {
    ex.skipped = true // terza volta: saltato per questa sessione
  }
  return s
}

/** Salta definitivamente l'esercizio corrente */
export function skipCurrent(session) {
  const s = structuredClone(session)
  const ex = getExercise(s, s.queue[0])
  ex.skipped = true
  s.queue = s.queue.slice(1)
  return s
}

/** Sposta un esercizio scelto manualmente in testa alla coda */
export function jumpTo(session, key) {
  const s = structuredClone(session)
  if (!s.queue.includes(key)) return s
  s.queue = [key, ...s.queue.filter((k) => k !== key)]
  return s
}

export function setNote(session, key, note) {
  const s = structuredClone(session)
  getExercise(s, key).note = note
  return s
}

export function togglePause(session) {
  const s = structuredClone(session)
  if (s.pauseStartedAt) {
    s.pausedMs += now() - s.pauseStartedAt
    s.pauseStartedAt = null
  } else {
    s.pauseStartedAt = now()
  }
  return s
}

export function finishSession(session) {
  const s = structuredClone(session)
  if (s.pauseStartedAt) {
    s.pausedMs += now() - s.pauseStartedAt
    s.pauseStartedAt = null
  }
  s.endedAt = now()
  const allDone = s.exercises.every((e) => e.skipped || e.series.every((x) => x.done))
  const anyDone = s.exercises.some((e) => e.series.some((x) => x.done))
  s.status = allDone || anyDone ? 'completed' : 'aborted'
  s.completedFully = allDone
  return s
}

/** Durata attiva in ms (esclude le pause) */
export function activeDuration(s, at = now()) {
  const end = s.endedAt || at
  let paused = s.pausedMs
  if (s.pauseStartedAt) paused += end - s.pauseStartedAt
  return Math.max(0, end - s.startedAt - paused)
}

/** Statistiche di riepilogo */
export function computeStats(s) {
  const totalSeries = s.exercises.reduce((n, e) => n + e.sets, 0)
  const doneSeries = s.exercises.reduce((n, e) => n + e.series.filter((x) => x.done).length, 0)
  const doneExercises = s.exercises.filter((e) => e.series.every((x) => x.done)).length
  const skipped = s.exercises.filter((e) => e.skipped).length

  const perExercise = s.exercises.map((e) => ({
    key: e.key,
    name: e.name,
    category: e.category,
    done: e.series.filter((x) => x.done).length,
    sets: e.sets,
    skipped: e.skipped,
    durationSec: e.startedAt && e.endedAt ? Math.round((e.endedAt - e.startedAt) / 1000) : null,
    avgRestSec: (() => {
      const rests = e.series.map((x) => x.restSec).filter((r) => r != null)
      return rests.length ? Math.round(rests.reduce((a, b) => a + b, 0) / rests.length) : null
    })(),
    note: e.note,
  }))

  const byCategory = {}
  for (const e of s.exercises) {
    const d = e.series.filter((x) => x.done).length
    if (d > 0) byCategory[e.category] = (byCategory[e.category] || 0) + d
  }

  const volumeKg = s.exercises.reduce(
    (v, e) =>
      v +
      e.series
        .filter((x) => x.done && x.actualWeightKg != null)
        .reduce((a, x) => a + x.actualReps * x.actualWeightKg, 0),
    0
  )

  const allRests = s.exercises.flatMap((e) => e.series.map((x) => x.restSec).filter((r) => r != null))
  const avgRestSec = allRests.length ? Math.round(allRests.reduce((a, b) => a + b, 0) / allRests.length) : null

  return {
    durationSec: Math.round(activeDuration(s) / 1000),
    startedAt: s.startedAt,
    endedAt: s.endedAt,
    totalSeries,
    doneSeries,
    doneExercises,
    totalExercises: s.exercises.length,
    skipped,
    perExercise,
    byCategory,
    volumeKg: Math.round(volumeKg),
    avgRestSec,
    restTargetSec: s.restDefaultSec,
    notes: s.exercises.filter((e) => e.note).map((e) => ({ name: e.name, note: e.note })),
  }
}
